// core/httpClient.js
/*
 *    Date: 
 *  Author: 
 * Project: 
 *
 * 
 */

var ModName = "HttpClient";

/* Modules */
var _https = require("https");
var _http = require("http");
var _log  = null;

/*
 * Public functions
 */
function init(log)
{
  _log = log;
}

/*
 * HTTP functions
 */
function get(opt,callback)
{
  if (typeof opt == "string")
    _log.info(ModName+" - exec GET "+opt);
  else
    _log.info(ModName+" - exec GET "+opt.host+":"+opt.port+opt.path);

  var req = _http.get(opt,function(res)
  {
    var resCode = res.statusCode;
    _log.info(ModName+" - response code: "+resCode);

    if (resCode >= 400)
      return callback(null);

    // Process response
    var body = "", bBodySent = false;

    res.on("error",function(err)
    {
      _log.error(ModName+" - response error: "+err.message);
      return callback(null);
    });

    res.on("data",function(chunk)
    {
      body += chunk;
    });

    res.on("end",function()
    {
      if (!bBodySent)
      {
        bBodySent = true;
        return callback(body);
      }
    });

    res.on("close",function()
    {
      if (!bBodySent)
      {
        bBodySent = true;
        return callback(body);
      }
    });
  });

  // Request error
  req.on("error",function(err)
  {
    _log.error(ModName+" - cannot exec GET: "+err.message);
    return callback(null);
  });
}

function getRaw(opt,callback)
{
  _log.info(ModName+" - exec GET "+opt.host+":"+opt.port+opt.path);

  var req = _http.get(opt,function(res)
  {
    var resCode = res.statusCode;
    _log.info(ModName+" - response code: "+resCode);

    if (resCode >= 400)
      return callback(null);

    // Process response
    var body = new Buffer(0), bBodySent = false;

    res.on("error",function(err)
    {
      _log.error(ModName+" - response error: "+err.message);
      return callback(null);
    });

    res.on("data",function(chunk)
    {
      body = Buffer.concat([body,chunk]);
    });

    res.on("end",function()
    {
      if (!bBodySent)
      {
        bBodySent = true;
        return callback(body);
      }
    });

    res.on("close",function()
    {
      if (!bBodySent)
      {
        bBodySent = true;
        return callback(body);
      }
    });
  });

  // Request error
  req.on("error",function(err)
  {
    _log.error(ModName+" - cannot exec GET: "+err.message);
    return callback(null);
  });
}

function post(opt,data,callback)
{
  _log.info(ModName+" - exec POST "+opt.host+":"+opt.port+opt.path);

  // Force method to POST
  opt.method = "POST";

  var req = _http.request(opt,function(res)
  {
    var resCode = res.statusCode;
    _log.info(ModName+" - response code: "+resCode);

    if (resCode >= 400)
      return callback(null);

    // Process response
    var body = "", bBodySent = false;

    res.on("error",function(err)
    {
      _log.error(ModName+" - response error: "+err.message);
      return callback(null);
    });

    res.on("data",function(chunk)
    {
      body += chunk;
    });

    res.on("end",function()
    {
      if (!bBodySent)
      {
        bBodySent = true;
        return callback(body);
      }
    });

    res.on("close",function()
    {
      if (!bBodySent)
      {
        bBodySent = true;
        return callback(body);
      }
    });
  });

  // Request error
  req.on("error",function(err)
  {
    _log.error(ModName+" - cannot exec POST: "+err.message);
    return callback(null);
  });

  // Post data
  req.end(data);
}

function postRaw(opt,data,callback)
{
  _log.info(ModName+" - exec POST "+opt.host+":"+opt.port+opt.path);

  // Force method to POST
  opt.method = "POST";

  var req = _http.request(opt,function(res)
  {
    var resCode = res.statusCode;
    _log.info(ModName+" - response code: "+resCode);

    if (resCode >= 400)
      return callback(null);

    // Process response
    var body = new Buffer(0), bBodySent = false;

    res.on("error",function(err)
    {
      _log.error(ModName+" - response error: "+err.message);
      return callback(null);
    });

    res.on("data",function(chunk)
    {
      body = Buffer.concat([body,chunk]);
    });

    res.on("end",function()
    {
      if (!bBodySent)
      {
        bBodySent = true;
        return callback(body);
      }
    });

    res.on("close",function()
    {
      if (!bBodySent)
      {
        bBodySent = true;
        return callback(body);
      }
    });
  });

  // Request error
  req.on("error",function(err)
  {
    _log.error(ModName+" - cannot exec POST: "+err.message);
    return callback(null);
  });

  // Post data
  req.end(data);
}

function put(opt,data,callback)
{
  _log.info(ModName+" - exec PUT "+opt.host+":"+opt.port+opt.path);

  // Force method to PUT
  opt.method = "PUT";

  var req = _http.request(opt,function(res)
  {
    var resCode = res.statusCode;
    _log.info(ModName+" - response code: "+resCode);

    if (resCode >= 400)
      return callback(null);

    // Process response
    var body = "", bBodySent = false;

    res.on("error",function(err)
    {
      _log.error(ModName+" - response error: "+err.message);
      return callback(null);
    });

    res.on("data",function(chunk)
    {
      body += chunk;
    });

    res.on("end",function()
    {
      if (!bBodySent)
      {
        bBodySent = true;
        return callback(body);
      }
    });

    res.on("close",function()
    {
      if (!bBodySent)
      {
        bBodySent = true;
        return callback(body);
      }
    });
  });

  // Request error
  req.on("error",function(err)
  {
    _log.error(ModName+" - cannot exec PUT: "+err.message);
    return callback(null);
  });

  // Send data
  req.end(data);
}

/*
 * HTTPS functions
 */
function req(opt,data,callback)
{
  _log.info(ModName+" - exec "+opt.method+
    " https://"+opt.host+":"+opt.port+opt.path);

  /*
   * Create request
   */
  var req = _https.request(opt,function(res)
  {
    var resCode = res.statusCode;

    /* Check response code */
    _log.info(ModName+" - response code: "+resCode);

    if (resCode >= 400)
    {
      callback(null);
      return;
    }

    /* Process response */
    var body = "";

    res.on("error",function(err)
    {
      _log.error(ModName+" - response error: "+err.message);
      callback(null);
    });

    res.on("data",function(chunk)
    {
      body += chunk;
    });

    res.on("end",function()
    {
      callback(body);
    });
  });

  req.on("error",function(err)
  {
    _log.error(ModName+" - cannot exec request: "+err.message);
    callback(null);
  });

  /*
   * Exec request
   */
  req.end(data);
}

/*
 * Exports
 */
exports.init = init;
exports.get = get;
exports.getRaw = getRaw;
exports.post = post;
exports.postRaw = postRaw;
exports.put = put;
exports.req = req;
