// cantieri/entity/civico.js
/*
 *    Date: 
 *  Author: 
 * Project: 
 *
 * 
 */

var entTemp = require(process.cwd()+"/../lib/entityTemplate").entityTemplate;
var util = require("util");
var self = {};
var config   = require("config/config.json");

var civico = function()
{
  civico.super_.call(this);

  this.moduleName = "Entity civico";
  this.entityName = "civico";
}

/*
 * Inheritance
 */
util.inherits(civico,entTemp);

civico.prototype.init = function(opt)
{
  self = this;
  civico.super_.prototype.init.call(self,opt);
}



civico.prototype.master = function(opt,callback)
{
  var reqOpt = {
    host: config.StreetorServer.host, // In config.json is setted host and port of streetor
    port: config.StreetorServer.port,
    path: "/civico/master",
    headers: {
      [config.Auth.cookieName]: config.Auth.innerToken, //opt.req.headers.it_app_auth,
      "content-type": "application/json"
    }
  };

  var reqBody = {
    ord:  "nome_via|ASC;numero|ASC;esponente|ASC",
    filter: opt.object.filter
  };

  self.httpClient.post(reqOpt,JSON.stringify(reqBody),function(resVia)
  {
    callback(null, resVia);
  });
}

/*
 * New methods
 */


/*
 * Export
 */
exports.civico = civico;

/*
 * Private function
 */
