// cantieri/entity/notice.js
/*
 *    Date: 
 *  Author: 
 * Project: 
 *
 * 
 */

var entTemp = require(process.cwd()+"/../lib/entityTemplate").entityTemplate;
var util = require("util");
var async = require("async");
var self = {};

var rlConfig   = require("config/rleConf.json");
var emConfig = require("config/emConf.json");

var notice = function()
{
  notice.super_.call(this);

  this.moduleName = "Entity notice";
  this.entityName = "notice";
}

/*
 * Inheritance
 */
util.inherits(notice,entTemp);

notice.prototype.init = function(opt)
{
  self = this;
  notice.super_.prototype.init.call(self,opt);
}

notice.prototype.master = function(opt,callback)
{
  var tasks = {};
  /*
  * Query to execute
  *
  * SELECT *
  * FROM view_roadsite / view_move
  * WHERE TO_CHAR(now(), 'DD/MM/YYYY') = to_char((date(start_date) + interval '-<interval> hour'), 'DD/MM/YYYY')
  * AND status_id = 3
  */

  var qStrR = "SELECT * " +
    "FROM view_roadsite vr " +
    "WHERE TO_CHAR(now(), 'DD/MM/YYYY') = to_char((date(start_date) + interval '-" +
    rlConfig.noticeInterval + " hour'), 'DD/MM/YYYY') " +
    "AND status_id = 3 AND (sended is false OR sended is null)";

  tasks['roadsite'] = function(cb){self.crud.select({queryString:qStrR}, [], cb);};


  var qStrM = "SELECT * " +
    "FROM view_move vm " +
    "WHERE TO_CHAR(now(), 'DD/MM/YYYY') = to_char((date(start_date) + interval '-" +
    rlConfig.noticeInterval + " hour'), 'DD/MM/YYYY') " +
    "AND status_id = 3 AND (sended is false OR sended is null)";

  tasks['move'] = function(cb){self.crud.select({queryString:qStrM}, [], cb);};

  // execute tasks
  async.parallelLimit(tasks,1,function(errT,resT)
  {
    var tasksN = {};
    if (resT['roadsite'] && resT['roadsite'].result && resT['roadsite'].result.length)
    {
      for (var idxR=0; idxR<resT['roadsite'].result.length; idxR++)
      {
        var itemR = resT['roadsite'].result[idxR];
        var noticeRoadsiteOpt =
        {
          id:itemR.id,
          towns: itemR.town,
          address: itemR.address,
          authority: itemR.authority,
          start_date: itemR.start_date
        }
        createNoticeTask(tasksN, idxR, noticeRoadsiteOpt, 'roadsite');
      }
    }
    if (resT['move'] && resT['move'].result && resT['move'].result.length)
    {

      for (var idxM=0; idxM<resT['move'].result.length; idxM++)
      {
        var itemM = resT['move'].result[idxM];
        var noticeMoveOpt =
        {
          id:itemM.id,
          towns: itemM.town,
          address: itemM.address,
          authority: itemM.authority,
          start_date: itemM.start_date
        }
        createNoticeTask(tasksN, idxM, noticeMoveOpt, 'move');
      }
    }
    async.parallelLimit(tasksN,1,function(errTNotice,resTNotice)
    {
      // console.log(resTNotice);
      callback(null, resTNotice);
    });

  });

}

/*
 * New methods
 */


/*
 * Export
 */
exports.notice = notice;

/*
 * Private function
 */
function createNoticeTask(tasksN,index,option,table)
{
  tasksN["notice-"+table+"-"+index] = function(cb){sendPushToApp(option,table, function(res)
    {
      cb(null,res);
    });};
}

function sendPushToApp(op,tableToUpdate,callback)
{
  // Retrieve id for given towns
  var id = op.id;
  var towns = op.towns;
  var address = op.address;
  var authority = op.authority;
  var start_date = op.start_date*1;

  var aTowns = [];
  queryVal = [];

  for (var idx = 0; idx < towns.split(";").length; idx++)
  {
    aTowns.push("'" + towns.split(";")[idx] + "'");
  }

  /*
   * Query to execute:
   *
   * SELECT *
   * FROM tows_ids
   * WHERE id = $1
   *
   */
  var queryOpt =
  {
    fields: self.crudUtils.ALL_FIELDS,
    from: [{name:"public.town_ids", type:self.crudUtils.TABLE}],
    where: [
      {
        typeCond: self.crudUtils.SIMPLE_COND,
        leftSide: "name",
        operator: self.crudUtils.IN,
        rightSide: aTowns.join()
      }
    ]
  };

  self.crud.select(queryOpt,queryVal,function(err,res)
  {
    if (err || !res)
    {
      self.log.error(self.moduleName + " - Error on retrieve ids for town " + towns);

      callback(err)
      return;
    }

    // Prepare post to send push
    var httpOpt = {
      host: emConfig.push.url,
      path: emConfig.push.path,
      port: 80,
      method: "POST",
      headers: {"Content-Type":"application/json"}
    };

    var date = new Date(start_date);

    var year = date.getFullYear(),
    mon = ("0" + (date.getMonth()+1)).slice(-2),
    day = ("0" + date.getDate()).slice(-2);

    var complete_date =  day + "/" + mon + "/" + year;

    var notice = {
      "id":id,
      "municipiIds": res.result.map(function(item){return item.id}),
      "title": "Apertura nuovo cantiere", // + authority,
      "message": "Il giorno " + complete_date + " sara' aperto un nuovo cantiere in " + address//authority
    }

    self.httpClient.post(httpOpt,JSON.stringify(notice),function(resPush)
    {
      self.log.info(self.moduleName + " - Response send push notice: "+resPush);

      if (resPush)
      {
        resPush = JSON.parse(resPush);

        // Update sended
        var qOpt = {
          table: {name:tableToUpdate, type:self.crudUtils.TABLE},
          fields: [{name:"sended"}],
          where: [{
            typeCond: self.crudUtils.SIMPLE_COND,
            leftSide: "id",
            operator: self.crudUtils.EQ,
            rightSide: "$2"
          }],
          returning: [
            {name: "id"}
          ]
        };

        qVal = [{value:true},{value:id}];

        self.crud.update(qOpt, qVal, function(errU,resU)
        {
          if(resU && resU.result && resU.result.length)
          {
            self.log.info(self.moduleName+" - Update sended attribute for roadsite with id " + id);
          }
          else
          {
            self.log.error(self.moduleName+" - Error on update sended attribute for roadsite with id " + id);
          }
          var retVal = resPush.resultCode == "OK" ? "OK" : "KO";
          callback({result:retVal});
        });
      }
      else
      {
        self.log.error(self.moduleName + " - Error on send push notice: "+resPush);
        callback({result:"KO"});
      }
    });

  });
}
