// cantieri/entity/via.js
/*
 *    Date: 
 *  Author: 
 * Project: 
 *
 * 
 */

var entTemp = require(process.cwd()+"/../lib/entityTemplate").entityTemplate;
var util = require("util");
var self = {};
var config   = require("config/config.json");

var via = function()
{
  via.super_.call(this);

  this.moduleName = "Entity via";
  this.entityName = "via";
}

/*
 * Inheritance
 */
util.inherits(via,entTemp);

via.prototype.init = function(opt)
{
  self = this;
  via.super_.prototype.init.call(self,opt);
}



via.prototype.master = function(opt,callback)
{
  var reqOpt = {
    host: config.StreetorServer.host, // In config.json is setted host and port of streetor
    port: config.StreetorServer.port,
    path: "/via/masterActive",
    headers: {
      [config.Auth.cookieName]: config.Auth.innerToken, //opt.req.headers.it_app_auth,
      "content-type": "application/json"
    }
  };

  var reqBody = {
    ord: "denominazione|ASC",
    filter: opt.object.filter
  };

  self.httpClient.post(reqOpt,JSON.stringify(reqBody),function(resVia)
  {
    callback(null, resVia);
  });
}

/*
 * New methods
 */


/*
 * Export
 */
exports.via = via;

/*
 * Private function
 */
